### PR TITLE
Polymorphism fixes

### PR DESCRIPTION
**Pull request by ghallak**
Fixes: #413

This introduces a couple of fixes for the referenced issue.

The commits should be merged without squashing.

### CODE REVIEW

File: src/aeso_ast_infer_types.erl [hunk 1]
```diff
@@ -909,55 +911,53 @@ infer1(Env, [{pragma, _, _} | Rest], Acc, Options) ->
...
+      Impls :: [aeso_syntax:con()],
+      DefinedContracts :: [aeso_syntax:decl()].
+populate_functions_to_implement(Env, ContractName, Impls, DefinedContracts) ->
+    create_type_errors(),
+    [ begin
+        Inserted = ets_insert_new(functions_to_implement, {name(Id), I, Decl}),
+        [{_, I2, _}] = ets_lookup(functions_to_implement, name(Id)),
+        Inserted orelse type_error({interface_implementation_conflict, ContractName, I, I2, Id})
+      end || {I, Decl = {fun_decl, _, Id, _}} <- functions_to_implement(Impls, DefinedContracts) ],
     destroy_and_report_type_errors(Env).
```
Comment (radrow) [inline src/aeso_ast_infer_types.erl:959]: If that throws, is the newly created ets table cleared?

File: src/aeso_ast_infer_types.erl [hunk 2]
```diff
@@ -1466,15 +1466,34 @@ check_reserved_entrypoints(Funs) ->
...
+    TypeSig = {type_sig, Ann, none, Named, Args, Ret},
+    delete_if_implementation(Env, Name, TypeSig),
+    {{Name, TypeSig}, {fun_decl, Ann, Id, Type1}};
 check_fundecl(Env, {fun_decl, Ann, Id = {id, _, Name}, Type}) ->
     type_error({fundecl_must_have_funtype, Ann, Id, Type}),
     {{Name, {type_sig, Ann, none, [], [], Type}}, check_type(Env, Type)}.
 
+%% Delete the function FunName with the signature FunSig from of functions
+%% to be implemented if it is included there, or return ok otherwise.
+-spec delete_if_implementation(env(), FunName, FunSig) -> true when
```
Comment (radrow) [inline src/aeso_ast_infer_types.erl:1478]: I don't like this name. It is not clear from where you delete (the universe?). At first I thought you delete it from the contract. Why not something like "register implementation"?

File: src/aeso_ast_infer_types.erl [hunk 3]
```diff
@@ -1466,15 +1466,34 @@ check_reserved_entrypoints(Funs) ->
...
+      FunName :: string(),
+      FunSig :: typesig().
+delete_if_implementation(Env, Name, Sig) ->
```
Comment (radrow) [inline src/aeso_ast_infer_types.erl:1481]: Also, why do you match type sig as well?

File: src/aeso_ast_infer_types.erl (hunk 1 above)
Comment (ghallak) [reply src/aeso_ast_infer_types.erl:959]: It is cleared inside `clean_up_ets`. In which case would we need to clean the `functions_to_implement` table explicitly?

File: src/aeso_ast_infer_types.erl (hunk 3 above)
Comment (ghallak) [reply src/aeso_ast_infer_types.erl]: Changed here 2a55823633625f40d7143d48157e0844a976769c

File: src/aeso_ast_infer_types.erl (hunk 2 above)
Comment (ghallak) [reply src/aeso_ast_infer_types.erl]: Changed here 06879fdf7384fff7c038196a2f5799aefdfcd98a

**Review by radrow — APPROVED** [review]

**Review by hanssv — APPROVED** [review]

**Pull request merged**